### PR TITLE
generate overloaded sigs for active record sum

### PR DESCRIPTION
**Pull request by stathis-alexander**
### Motivation
Closes https://github.com/Shopify/tapioca/issues/1822. See discussion there for full context. 

TL;DR: The current sigs are incorrect when given a block.

### Tests
I believe I updated the spec here correctly.

### CODE REVIEW

File: spec/tapioca/dsl/compilers/active_record_relations_spec.rb [hunk 1]
```diff
@@ -229,8 +229,9 @@ def second_to_last!; end
                     def sole; end
 
                 <% end %>
-                    sig { params(column_name: T.nilable(T.any(String, Symbol)), block: T.nilable(T.proc.params(record: T.untyped).returns(T.untyped))).returns(Numeric) }
-                    def sum(column_name = nil, &block); end
+                    sig { params(initial_value_or_column: T.nilable(T.any(String, Symbol))).returns(Numeric) }
+                    sig { type_parameters(:U).params(initial_value_or_column: T.nilable(T.type_parameter(:U)), block: T.proc.params(object: ::Post).returns(T.type_parameter(:U)).returns(T.type_parameter(:U)) }
```
[the remark this answers is not in the document]
Comment (stathis-alexander) [reply spec/tapioca/dsl/compilers/active_record_relations_spec.rb:233]: Great catch, thanks!

**Comment by stathis-alexander** [comment]
Is it possible to have someone add a label to and review this PR? I don't seem to be able to.

File: lib/tapioca/dsl/compilers/active_record_relations.rb [hunk 2]
```diff
@@ -788,13 +788,34 @@ def create_common_methods
...
               )
             when :sum
-              create_common_method(
-                "sum",
+              sigs = [
+                common_relation_methods_module.create_sig(
+                  parameters: [create_opt_param(
+                    "initial_value_or_column",
+                    type: "T.nilable(T.any(String, Symbol))",
+                    default: "0",
```
[the remark this answers is not in the document]
Comment (stathis-alexander) [reply lib/tapioca/dsl/compilers/active_record_relations.rb:796]: Hi @egiurleo. Great question! Maybe there is a better way to define these sigs. Maybe just remove this default value?

For reference - [here is the definition of this method in active record](https://github.com/rails/rails/blob/89adf24a570b0adf9f62011339b042046471f421/activerecord/lib/active_record/relation/calculations.rb#L172). 

The default value here is indeed `0`. You can see that there are two different forms this method can take (hence the overloaded signatures here):
1. a column name with no block
2. an initial value with a block
In the case no column name or initial value is provided, it assumes it is the later. 

Should I drop the default value from the first sig? How would you write these? In my (admittedly brief) testing, this seemed to work.

[the remark this answers is not in the document]
Comment (paracycle) [reply lib/tapioca/dsl/compilers/active_record_relations.rb:796]: The value of `initial_value_or_column` can be anything that is `to_s`able, as can be seen by:
```irb
irb(main):012> Model.sum
  Model Sum (0.7ms)  SELECT SUM(0) FROM `models`
=> 0
irb(main):016> Model.sum(100)
  Model Sum (0.5ms)  SELECT SUM(100) FROM `models`
=> 3300
irb(main):013> class Foo; def to_s; "id"; end; end
=> :to_s
irb(main):015> Model.sum(Foo.new)
  Model Sum (0.6ms)  SELECT SUM(`services`.`id`) FROM `models`
=> 2415
```

The signatures we want are:
```ruby
sig { params(initial_value_or_column: T.untyped).returns(Numeric) }
sig { type_parameters(:U).params(initial_value_or_column: T.nilable(T.type_parameter(:U)), block: T.proc.params(object: Constant).returns(T.type_parameter(:U))).returns(T.type_parameter(:U)) }
def sum(initial_value_or_column = T.unsafe(nil), &block); end
```

[the remark this answers is not in the document]
Comment (stathis-alexander) [reply lib/tapioca/dsl/compilers/active_record_relations.rb]: @paracycle  - That makes sense to me. I've updated the PR. Mind taking a look?

**Comment by bdewater** [comment]
https://github.com/Shopify/tapioca/pull/1856 was merged, this needs a rebase.

**Comment by stathis-alexander** [comment]
@bdewater Thanks, I've rebased with `main`.

**Comment by bdewater** [comment]
Tapioca monkey patches  the rbi gem to add `create_sig` so this method ends up in `sorbet/rbi/gems/rbi@0.1.10.rbi`. In the current state the patch signature and the RBI file signature don't match and Sorbet complains. Running `bundle exec tapioca gem rbi` to force it to recompile the RBI file for rbi (the gem) should fix the issue 😄

**Comment by KaanOzkan** [comment]
@stathis-alexander can you rebase on latest `main`?

**Comment by stathis-alexander** [comment]
@KaanOzkan  - Done!

File: lib/tapioca/dsl/compilers/active_record_relations.rb [hunk 3]
```diff
@@ -822,13 +822,28 @@ def create_common_methods
...
+                  return_type: "T.type_parameter(:U)",
+                ),
+              ]
+              common_relation_methods_module.create_method_with_sigs(
+                method_name.to_s,
+                sigs: sigs,
                 parameters: [
-                  create_opt_param("column_name", type: "T.nilable(T.any(String, Symbol))", default: "nil"),
-                  create_block_param("block", type: "T.nilable(T.proc.params(record: T.untyped).returns(T.untyped))"),
+                  RBI::OptParam.new("initial_value_or_column", "T.unsafe(nil)"),
```
Comment (KaanOzkan) [inline lib/tapioca/dsl/compilers/active_record_relations.rb:844]: Do we need the `T.unsafe()` here?

Comment (stathis-alexander) [reply lib/tapioca/dsl/compilers/active_record_relations.rb]: @KaanOzkan  - updated!

Comment (KaanOzkan) [reply lib/tapioca/dsl/compilers/active_record_relations.rb]: Thanks it looked fine on my tests.

**Review by egiurleo — APPROVED** [review]

**Pull request merged**